xapian-applications/omega/handler_gmime.cc

```diff
@@ -108,39 +109,28 @@ parser_content(GMimeObject* me, string& dump)
 	GMimeDataWrapper* content = g_mime_part_get_content_object(part);
 	string type = g_mime_content_type_get_media_type(ct);
 	string subtype = g_mime_content_type_get_media_subtype(ct);
-	string charset(g_mime_object_get_headers(me));
-	size_t end, start;
-	start = charset.find("charset=");
-	if (start != string::npos) {
-	    start += 8;
-	    if (charset[start] == '"')
-		start++;
-	    for (end = start; end < charset.length() &&
-		 !isspace(charset[end]) && charset[end] != '"'; ++end);
-	    charset = charset.substr(start, end - start);
-	    charset = g_mime_charset_canon_name(charset.c_str());
-	} else {
-	    charset.clear();
-	}
+	string charset;
+	const char* p = g_mime_object_get_content_type_parameter(me, "charset");
+	if (p) charset = g_mime_charset_canon_name(p);
 	if (type == "text") {
+	    string text;
 	    char buffer[SIZE];
 	    GMimeStream* sr = g_mime_data_wrapper_get_stream(content);
-	    size_t len = g_mime_stream_read(sr, buffer, SIZE);
 	    auto encoding = g_mime_part_get_content_encoding(part);
-	    len = decode((unsigned char*)buffer, len, encoding);
-	    if (subtype == "plain") {
-		if (!charset.empty() && charset != "UTF-8") {
-		    GMimeFilter* ch = g_mime_filter_charset_new(charset.c_str(),
-								"UTF-8");
-		    if (GMIME_IS_FILTER(ch)) {
-			size_t presize = 0;
-			g_mime_filter_complete(ch, buffer, len, presize,
-					      (char**)&buffer, &len, &presize);
-		    }
+	    int len = 0;
+	    do {
+		len = g_mime_stream_read(sr, buffer, SIZE / 2);
+		if (len > 0) {
+		    len = decode((unsigned char*)buffer, len, encoding);
```

Comment:
Quoting the dev guide:

> Use C++ style casts (``static_cast<>``, ``reinterpret_cast<>``, and
``const_cast<>``) or constructor-syntax (e.g. ``double(value)``) in
preference to C style casts.  The syntax of the C++ casts is ugly,
but they do make the intent much clearer which is definitely a good
thing, and they avoid issues such as casting away const when you
only meant to cast the type of a pointer.